Cgen uses wrong type

### Describe the bug

<title>

### Reproduction Steps

```v
pub type EventListener[T] = fn (T) !
pub type Check[T] = fn (T) bool

pub struct EventController[T] {
mut:
	id        int
	listeners map[int]EventListener[T]
}

fn (mut ec EventController[T]) generate_id() int {
	return ec.id++
}

@[params]
pub struct EmitOptions {
pub:
	error_handler ?fn (int, IError)
}

pub fn (mut ec EventController[T]) emit(e T, options EmitOptions) {
	if ec.listeners.len == 1 {
		f := (ec.listeners.values()[0])
		f(e) or {
			if g := options.error_handler {
				g(0, err)
			}
		}
		return
	}
}

struct Foo {}
struct Bar {}

fn main() {
	foo := EventController[Foo]{}
	bar := EventController[Bar]{}
}
```

### Expected Behavior

Successful compile

### Current Behavior

```
test.v:36:2: warning: unused variable: `foo`
   34 |
   35 | fn main() {
   36 |     foo := EventController[Foo]{}
      |     ~~~
   37 |     bar := EventController[Bar]{}
   38 | }
test.v:37:2: warning: unused variable: `bar`
   35 | fn main() {
   36 |     foo := EventController[Foo]{}
   37 |     bar := EventController[Bar]{}
      |     ~~~
   38 | }
==================
C:/Users/mclr/AppData/Local/Temp/v_0/test.16401507672092969587.tmp.c:6877: warning: implicit declaration of function 'tcc_backtrace'
C:/Users/mclr/AppData/Local/Temp/v_0/test.16401507672092969587.tmp.c:12878: error: cannot convert 'struct main__Foo' to 'struct main__Bar'
...
==================
(Use `v -cg` to print the entire error message)

builder error:
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.4.3 5306469

### Environment details (OS name and version, etc.)

```
V full version: V 0.4.3 23ed1e9.5306469
OS: windows, Microsoft Windows 11 Pro v22000 64-bit
Processor: 12 cpus, 64bit, little endian, 

getwd: D:\Games\Proekti\V\interactions
vexe: D:\Games\Proekti\V\v\v.exe
vexe mtime: 2023-12-20 16:23:28

vroot: OK, value: D:\Games\Proekti\V\v
VMODULES: OK, value: C:\Users\mclr\.vmodules
VTMP: OK, value: C:\Users\mclr\AppData\Local\Temp\v_0

Git version: git version 2.37.0.windows.1
Git vroot status: weekly.2023.51-13-g53064697
.git/config present: true

CC version:
thirdparty/tcc status: thirdparty-windows-amd64 a39eb79b
```
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/checker/fn.v
+++ b/vlib/v/checker/fn.v
@@ -1741,10 +1741,10 @@ fn (mut c Checker) method_call(mut node ast.CallExpr) ast.Type {
 	} else if (left_sym.kind == .map || final_left_sym.kind == .map)
 		&& method_name in ['clone', 'keys', 'values', 'move', 'delete'] {
 		if left_sym.kind == .map {
-			return c.map_builtin_method_call(mut node, unwrapped_left_type, c.table.sym(unwrapped_left_type))
+			return c.map_builtin_method_call(mut node, left_type)
 		} else if left_sym.info is ast.Alias {
-			parent_type := c.unwrap_generic(left_sym.info.parent_type)
-			return c.map_builtin_method_call(mut node, parent_type, c.table.final_sym(unwrapped_left_type))
+			parent_type := left_sym.info.parent_type
+			return c.map_builtin_method_call(mut node, parent_type)
 		}
 	} else if left_sym.kind == .array && method_name in ['insert', 'prepend'] {
 		if method_name == 'insert' {
@@ -2652,9 +2652,16 @@ fn (mut c Checker) check_map_and_filter(is_map bool, elem_typ ast.Type, node ast
 	}
 }
 
-fn (mut c Checker) map_builtin_method_call(mut node ast.CallExpr, left_type ast.Type, left_sym ast.TypeSymbol) ast.Type {
+fn (mut c Checker) map_builtin_method_call(mut node ast.CallExpr, left_type_ ast.Type) ast.Type {
 	method_name := node.name
 	mut ret_type := ast.void_type
+	// resolve T
+	left_type := if c.table.final_sym(left_type_).kind == .any {
+		c.unwrap_generic(left_type_)
+	} else {
+		left_type_
+	}
+	left_sym := c.table.final_sym(left_type)
 	match method_name {
 		'clone', 'move' {
 			if node.args.len != 0 {

--- a/vlib/v/tests/generics_array_method_call_with_multi_types_test.v
+++ b/vlib/v/tests/generics_array_method_call_with_multi_types_test.v
@@ -1,0 +1,44 @@
+pub type EventListener[T] = fn (T) !
+
+pub type Check[T] = fn (T) bool
+
+pub struct EventController[T] {
+mut:
+	id        int
+	listeners map[int]EventListener[T]
+}
+
+fn (mut ec EventController[T]) generate_id() int {
+	return ec.id++
+}
+
+@[params]
+pub struct EmitOptions {
+pub:
+	error_handler ?fn (int, IError)
+}
+
+pub fn (mut ec EventController[T]) emit(e T, options EmitOptions) {
+	if ec.listeners.len == 1 {
+		f := ec.listeners.values()[0]
+		f(e) or {
+			if g := options.error_handler {
+				g(0, err)
+			}
+		}
+		return
+	}
+}
+
+struct Foo {}
+
+struct Bar {}
+
+fn test_generic_array_method_call_with_multi_types() {
+	foo := EventController[Foo]{}
+	println(foo)
+	assert foo.id == 0
+	bar := EventController[Bar]{}
+	println(bar)
+	assert bar.id == 0
+}

